### PR TITLE
Password auth and role management

### PR DESCRIPTION
**Pull request by connor4312**
This PR adds role/user management and password authentication to the client. It was rather rough going (hence the big delay here!) mostly due to lacking documentation both in etcd and grpc; had to reverse engineer the Go client and some GRPC internals.

The diff is big but not quite as big as it looks -- I include some certs and keys for testing (grpc refuses to use any kind of authentication when not using TLS) which Github wants to diff. Protobuf generation logic was also changed to fix issues with mismatching and normalize casing to be more JavaScripty. Changes are  pretty contained:
 - src/types/grpc.d.ts are the newly-added types. Most of the grpc API, but not all of it
 - src/connection-pool.ts are the gritty changes to handle authing with grpc
 - src/auth.ts is the higher level builders for role and user management. Nothing crazy there, just pushing some bits around.

Remaining todo is to handle expiration / loss of authentication keys and transparently re-request them when that occurs. Will be in a followup PR.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -49,7 +50,7 @@
     "tslint": "^4.0.0",
     "tslint-microsoft-contrib": "4.0.0",
     "typedoc": "^0.5.10",
-    "typescript": "^2.2.2"
+    "typescript": "2.3.0"
```
Comment (SimonSchick) [inline package.json]: Did you mean ^?

File: proto/auth.proto [hunk 2]
```diff
@@ -10,9 +10,9 @@ message User {
 // Permission is a single entity
 message Permission {
   enum Type {
-    READ = 0;
-    WRITE = 1;
-    READWRITE = 2;
+    read = 0;
```
Comment (SimonSchick) [inline proto/auth.proto]: `Read`.

File: proto/kv.proto [hunk 3]
```diff
@@ -21,8 +21,8 @@ message KeyValue {
 }
 message Event {
   enum EventType {
-    PUT = 0;
-    DELETE = 1;
+    put = 0;
```
Comment (SimonSchick) [inline proto/kv.proto]: `Put`

File: proto/rpc.proto [hunk 4]
```diff
@@ -292,16 +292,16 @@ message ResponseHeader {
 }
 message RangeRequest {
   enum SortOrder {
-	NONE = 0; // default, no sorting
-	ASCEND = 1; // lowest target value first
-	DESCEND = 2; // highest target value first
+	none = 0; // default, no sorting
```
Comment (SimonSchick) [inline proto/rpc.proto]: See above

File: src/errors.ts [hunk 5]
```diff
@@ -78,6 +118,13 @@ const grpcMessageToError = new Map<string | RegExp, IErrorCtor>([
   ['Cancelled before creating subchannel', GRPCCancelledError],
   ['Pick cancelled', GRPCCancelledError],
   ['Disconnected', GRPCCancelledError],
+  [/role name already exists/, EtcdRoleExistsError],
```
Comment (SimonSchick) [inline src/errors.ts]: Why do these use RegExp's rather than strings like above?

File: src/index.ts [hunk 6]
```diff
@@ -90,12 +94,56 @@ export class Etcd3 {
...
+    value: string | Buffer | number,
+  ): Builder.ComparatorBuilder {
     return new Builder.ComparatorBuilder(this.kv).and(key, column, cmp, value);
   }
 
   /**
+   * Creates a structure representing an etcd range. Used in permission grants
+   * and queries. This is a convenience method for `Etcd3.Range.from(...)`.
+   */
+  public range(r: rangable): Range {
```
Comment (SimonSchick) [inline src/index.ts]: Types should be capitalised.

File: src/rpc.ts [hunk 7]
```diff
@@ -76,7 +74,7 @@ export class WatchClient {
    * last compaction revision.
    */
   public watch(): Promise<IDuplexStream<IWatchRequest, IWatchResponse>> {
-    return this.client.getConnection('Watch').then(cnx => cnx.watch());
+    return this.client.getConnection('Watch').then(cnx => (<any> cnx.client).watch());
```
Comment (SimonSchick) [inline src/rpc.ts]: Why the any cast? Missing typings? Tried adding patching them in via interface augmentation?

File: src/types/grpc.d.ts [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+   * The Service associated with the client, used for creating GRPC servers.
+   */
+  service: Service;
+}
+
+export class Server {
+  /**
+   * Creates a new GRPC server.
+   */
+  constructor();
```
Comment (SimonSchick) [inline src/types/grpc.d.ts]: rm?

File: src/types/grpc.d.ts [hunk 9]
```diff
@@ -0,0 +1,218 @@
...
+export function waitForClientRead(client: Client, deadline: Date | Number, callback: (err: Error | null) => void): void;
+
+/**
+ * Class for storing metadata. Keys are normalized to lowercase ASCII.
+ */
+export class Metadata {
+  /**
+   * Adds the given value for the given key. Normalizes the key.
+   */
+  add(key: string, value: string | Buffer): void;
```
Comment (SimonSchick) [inline src/types/grpc.d.ts]: public/private?

File: test/client.test.ts [hunk 10]
```diff
@@ -3,20 +3,43 @@ import * as sinon from 'sinon';
...
 } from '../src';
-import { getHosts } from './util';
+import { getOptions } from './util';
+
+function expectReject(promise: Promise<any>, err: { new (message: string): Error }) {
+  return promise
+    .then(() => { throw new Error('expected to reject'); })
+    .catch(actualErr => {
+      if (!(actualErr instanceof err)) {
+        console.error(actualErr.stack);
```
Comment (SimonSchick) [inline test/client.test.ts]: rm debug

File: test/util.ts [hunk 11]
```diff
@@ -1,6 +1,23 @@
...
 }
+
+/**
+ * Returns etcd options to use for connections.
+ */
+export function getOptions(defaults: Partial<IOptions> = {}): IOptions {
+  return {
+    hosts: getHost(),
+    credentials: { rootCertificate },
+    ...defaults,
```
Comment (SimonSchick) [inline test/util.ts]: Wouldn't the defaults override in this case?

File: package.json (hunk 1 above)
Comment (connor4312) [inline package.json]: I didn't. It looks like TS 2.3.1 has issues with tslint.

Comment (SimonSchick) [inline package.json]: Do you still use tslint 4 with ts 2.3? You are lucky this works at all lol.

File: src/rpc.ts (hunk 7 above)
Comment (connor4312) [inline src/rpc.ts]: This is a limitation of TypeScript+protobuf unfortunately. It's a [client](https://github.com/mixer/etcd3/blob/password-auth/src/types/grpc.d.ts#L32) class which has methods beyond what we provide by default which are not known at compile time.

File: test/util.ts (hunk 11 above)
Comment (connor4312) [inline test/util.ts]: Yep! That's intended. In some cases we purposely pass invalid hosts in and want them to override.

File: test/client.test.ts (hunk 10 above)
Comment (connor4312) [inline test/client.test.ts]: I actually do intend this here. The Chai instanceof check failing only prints the error message, it's useful to print the stack too to help diagnose issues (similarly to how we dump responses entirely on failed assertions in the backend integration tests).

File: proto/auth.proto (hunk 2 above)
Comment (connor4312) [inline proto/auth.proto]: Erp, somehow that totally slipped my mind

File: package.json (hunk 1 above)
Comment (connor4312) [inline package.json]: Yea. I imported the linting config we use in other ts, which is all tslint 4. We need to update soon, but that involves updating tslint-contrib-microsoft too, which is ehh. For a followup PR!

**Pull request merged**